### PR TITLE
Feat: add separate slot feature for spine animation component.

### PR DESCRIPTION
**Pull request by johanzhu**
- Add separate slot feature for spine animation component. User can separate slot independent. 
MeshGenerator will add sub mesh when adding separate slot. 
- Add hackSeparateSlotTexture function. User can change texture of separate slot.

### CODE REVIEW

File: src/core/MeshGenerator.ts [hunk 1]
```diff
@@ -73,14 +72,18 @@ export class MeshGenerator {
...
     const drawOrder = skeleton.drawOrder;
     const clipper = this.clipper;
+    const seperateSubMesh = [];
+    const restSubMesh = [];
     let vertices: ArrayLike<number> = this.vertices;
     let triangles: Array<number> = null;
     let uvs: ArrayLike<number> = null;
+    let start = 0;
+    let count = 0;
     for (let slotIndex = 0; slotIndex < drawOrder.length; slotIndex += 1) {
```
Comment (singlecoder) [inline src/core/MeshGenerator.ts:87]: drawOrder.length need cache

File: src/core/MeshGenerator.ts [hunk 2]
```diff
@@ -189,22 +192,60 @@ export class MeshGenerator {
...
+          }
+          start = indicesLength + finalIndicesLength;
+        } else {
+          count += finalIndicesLength;
+        }
+        
+        indicesLength += finalIndicesLength;
+
+        const materials = meshRenderer.getMaterials();
+        for (let i = 0; i < materials.length; i += 1) {
```
Comment (singlecoder) [inline src/core/MeshGenerator.ts:222]: materials.length need cache

File: src/core/MeshGenerator.ts [hunk 3]
```diff
@@ -189,22 +192,60 @@ export class MeshGenerator {
...
+    if (count > 0) {
+      const subMesh = new SubMesh(start, count);
+      restSubMesh.push(subMesh);
+      count = 0;
+    }
+
+    // update subMesh
+    this.spineMesh.mesh.clearSubMesh();
+    const subMeshes = seperateSubMesh.concat(restSubMesh);
+    for (let i = 0; i < subMeshes.length; i += 1) {
```
Comment (singlecoder) [inline src/core/MeshGenerator.ts:246]: subMeshes.length need cache

File: src/core/MeshGenerator.ts [hunk 4]
```diff
@@ -189,22 +192,60 @@ export class MeshGenerator {
...
+      this.spineMesh.mesh.addSubMesh(subMeshes[i]);
```
Comment (singlecoder) [inline src/core/MeshGenerator.ts:247]: this.spineMesh.mesh need cache

File: example/App.jsx [hunk 5]
```diff
@@ -33,7 +33,7 @@ function App() {
       const camera = cameraEntity.addComponent(o3.Camera);
       camera.farClipPlane = 2000000;
       camera.nearClipPlane = 0.001;
-      cameraEntity.transform.position = new o3.Vector3(0, 0, 12);
+      cameraEntity.transform.position = new o3.Vector3(0, 0, 120);
```
Comment (gz65555) [inline example/App.jsx:36]: use `setPosition` instead

File: src/SpineAnimation.ts [hunk 6]
```diff
@@ -60,6 +62,44 @@ export class SpineAnimation extends Script {
     this._meshGenerator.initialize(this._skeleton);
   }
 
+  addSeparateSlot(slotName: string) {
```
Comment (GuoLei1990) [inline src/SpineAnimation.ts:65]: Add comments.

File: src/SpineAnimation.ts [hunk 7]
```diff
@@ -60,6 +62,44 @@ export class SpineAnimation extends Script {
...
+      // add default material for new sub mesh
+      // split will generate two material
+      meshRenderer.setMaterial(materialCount, mtl);
+      meshRenderer.setMaterial(materialCount + 1, mtl);
+    } else {
+      console.warn(`Slot: ${slotName} not find.`);
+    }
+  }
+
+  hackSeparateSlotTexture(slotName: string, texture: Texture2D) {
```
Comment (GuoLei1990) [inline src/SpineAnimation.ts:87]: Add comments.

File: src/core/MeshGenerator.ts [hunk 8]
```diff
@@ -73,15 +72,21 @@ export class MeshGenerator {
...
+    let verticesLength = 0;
+    let indicesLength = 0;
 
     const drawOrder = skeleton.drawOrder;
-    const clipper = this.clipper;
+    const maxSlotCount = drawOrder.length;
+    const { clipper, spineMesh } = this;
+    const { mesh } = spineMesh;
+    const seperateSubMesh = [];
+    const restSubMesh = [];
```
Comment (GuoLei1990) [inline src/core/MeshGenerator.ts:83]: Will cause  GC every farme

**Review by GuoLei1990 — APPROVED** [review]
+1

**Pull request merged**